### ISSUE TITLE
Feature Request: RFC 8766 Discovery Proxy for Multicast DNS-Based Service Discovery

### ISSUE THREAD

**Issue by misch42** [avahi/avahi#334]
RFC8766 describes a Discovery Proxy that allows remote clients to query for hosts and and services on a local network. Any plans to implement such a Proxy into avahi?

Or are you aware of any other software, besides apple, that implemented this Proxy already?

**Comment by pusateri** [comment]
[Discovery Proxy](https://www.rfc-editor.org/rfc/rfc8766.html) also requires [RFC 8765](https://www.rfc-editor.org/rfc/rfc8765.html) (DNS Push Notifications) which depends on [RFC 8490](https://www.rfc-editor.org/rfc/rfc8490.html) (DNS Stateful Operations). The original experimental work for Discovery Proxy was done in an open source version of mDNSResponder. The repository is https://github.com/IETF-Hackathon/mDNSResponder

A Discovery Proxy is essentially a full authoritative server for one or more subdomains combined with the browse features of mDNS. It requires delegation from the parent domain for each subdomain on each .local subnet the proxy is operating on.

This technology is really nice but it's a big project.

**Comment by evverx** [comment]
> Any plans to implement such a Proxy into avahi?

I'm afraid not.

It should probably be possible to embed the low-level avahi-core API into a separate project to cover the mDNS part (to judge from https://github.com/avahi/avahi/issues/344 there was at least one attempt to do that using the avahi-client API  but it's too high-level for that). The unicast DNS part is a big project as mentioned and unlikely to be implemented here. Sorry.

**Comment by pemensik** [comment]
I admit ideas to have spaces in unicast names are quite unusual for DNS folks. Something like ``My Printer._ipp._tcp.Building 1.example.com.``, that is super weird. We may support spaces just in the first label and nowhere else. But anyway, we won't have this soon.

**Comment by footlooseboss** [comment]
TL;DR

> ...
> Discovery Proxy should not restrict to a specific subdomain format for better user experience.
>
> In examples, spaces are shown as actual spaces, but if manually entered into a zone file, represent spaces as '\032'. For instance, "My Printer._ipp._tcp.Building 1.example.com" becomes "My\032Printer._ipp._tcp.Building\0321.example.com" in a textual zone file.
>
> Note: '\032' doesn't appear in DNS messages over the air. In DNS wire format, spaces are sent as spaces, not as '\032'. Similarly, in a graphical user interface, spaces are shown as spaces, not as '\032'.
> ...

**Comment by footlooseboss** [comment]
[IETF-Hackathon/mDNSResponder](https://github.com/IETF-Hackathon/mDNSResponder)

**Comment by evverx** [comment]
> this would be nice to have

It would be nice to actually cover at least the mDNS RFC first.

> this would be very interesting feature

It doesn't matter whether it's an interesting feature or not. It's a huge engineering effort (in terms of testing, maintenance and so on).

> it might be possible

To be honest it's overly optimistic and I'm not even sure where it comes from.

> someone would be actively working on this

If a PR with this feature was opened now how long do you think it would take for anyone to even review it?

**Comment by footlooseboss** [comment]
[Cisco DNA Service for Bonjour - Traditional Networks Service Routing Overview](http://www.youtube.com/watch?v=qtRcb00Ue7E&t=218.76842222222223)

**Comment by pemensik** [comment]
It seems to me this were merged after hackaton into [Apple's mDNSResponder](https://github.com/apple-oss-distributions/mDNSResponder), which is used on apple products as a base system resolver. That implements both mDNS and DNS services if I understood it correctly. That essentially replaces whole Avahi daemon.

**Comment by evverx** [comment]
Once again I don't think it's feasible to bring it to avahi. It should focus on mDNS and whoever actually has the resources to maintain that sort of thing can utilize it in their projects like https://github.com/openthread/ot-br-posix/pull/868 for example. To be honest I'm not sure what the point of this theoretical discussion is.

**Comment by dgtlrift** [comment]
How does [RFC 8766](https://datatracker.ietf.org/doc/rfc8766/) differentiate from the proposals within IETF [DNS-SD Working Group](https://datatracker.ietf.org/wg/dnssd) and specifically the IETF Proposed Standard draft [Advertising Proxy for DNS-SD Service Registration Protocol](https://datatracker.ietf.org/doc/draft-ietf-dnssd-advertising-proxy/) and does it make sense to consider the direction that the DNS-SD Working Group is taking the standard?

**Comment by evverx** [comment]
I'll go ahead and close it as no-planned.

The openthread project where border routers should support https://www.rfc-editor.org/rfc/rfc8766.html can be used instead.
